docs/plugins/testing_workshop_docs/Pytest-testing-frameworks.md

```diff
@@ -0,0 +1,104 @@
+    def test_get_grade_fail():  
+        grade = get_grade_from_mark(65)  
+    >   assert grade == “Fail”, f”Expected 65 to fail but result was {grade}”  
+    E   AssertionError: Expected 65 to fail but result was Pass  
+    E   assert ‘Pass’ == ‘Fail’  
+    E     - Fail  
+    E     + Pass  
+     
+    example_func_py: 16: AssertionError  
+    ======================== short test summary info =========================  
+    FAILED example_func_py::test_get_grade_fail - AssertionError: Expected 65 to fail, but result was Pass  
+    ====================== 1 failed in 0.56s =======================  
+    (napari-env) user@directory %   
+
+## Parametrization  
+Another very useful tool pytest provides is parametrization.  
+    
+We've tested these functions with a single value. We need to be more thorough. Pytest allows us to parametrize tests. We decorate our function with `@pytest.mark.parametrize` and pass the decorator a parameter name, mark, as a string, and a list of values for which we’d like to run the test function. Note that we pass in 50 as an edge case; it's the lowest mark that will pass.  
+    
+    def get_grade_from_mark(mark):
+        if mark > 50
+            return “Pass”
+        else: 
+            return “Fail”
+ 
+    @pytest.mark.parametrize(‘mark’, [65, 80,50])
+    def test_get_grade_pass(mark):
+        grade = get_grade_from_mark(mark):  
+        assert grade == “Pass”, f”Expected {mark} to pass, but result was{grade}”
+ 
+    def test_get_grade_fail(mark):
+        grade = get_grade_from_mark(mark):
+        assert grade == “Fail”, f”Expected {mark} to fail, but result was{grade}”
+    
+        if _name_ == “_main_”;
+            test_get_grade_pass()
+            test_get_grade_fail()
+            print(“All passing.”)
+
+We run `pytest` which finds and runs `test_get_grade_pass(mark)`. `test_get_grade_pass(mark)` fails because the code says `mark > 50`, so a mark of 50 still fails. The code should say `>=50` for a mark to pass.  
+   
+    example_func.py .FF  
+    ================================= FAILURES ===============================  
+    —------------------------ test_get_grade_pass[50] ------------------------  
+    mark = 50  
+    @pytest parametrize(‘mark’, [65, 80, 50])  
+    def test_get_grade_pass(mark):  
+        grade = get_grade_from_mark(mark)  
+    >   assert grade == “Pass”, f”Expected {mark} to pass, but result was {grade}”  
+    E   AssertionError: Expected 50 to pass but result was fail  
+    E   assert ‘Fail’ == ‘Pass’  
+    E     - Pass  
+    E     + Fail  
+ 
+    example_func_py:12: AssertionError  
+    ======================== short test summary info =========================  
+    FAILED example_func_py::test_get_grade_pass[50] - AssertionError: Expected 50 to pass, but result was Fail   
+    
+Another valuable feature of `pytest` is the `pytest-cov` option discussed in the [Test coverage](./Test-Coverage.md) article.  
```

Comment:
In general, it is a good idea to not use capitalization in file names, to avoid problems in different operational systems which treat capitalization differently. In this case, for example, this file is not found when building because the casing used here is different than the filename. I'd suggest using lowercase only for all filenames, including extensions (please note you have some .PNG files which can also create errors when building in case-sensitive systems)

```suggestion
Another valuable feature of `pytest` is the `pytest-cov` option discussed in the [Test coverage](./Test-coverage.md) article.  
```